src/manager.S

```diff
@@ -0,0 +1,169 @@
+ *
+ * @return Pointer to newly allocated memory region
+ */
+.type init_compartments, "function"
+init_compartments:
+    // Compute size of required memory, equivalent to `length` parameter of
+    // `mmap`
+    mov       x0, #COMP_SIZE
+    mov       x1, #MAX_COMP_COUNT
+    mov       x2, #32    // size of the 2 switcher capabilities
+    madd      x1, x0, x1, x2
+
+    // Store length and `lr` on stack, as we'll need them later
+    stp       x1, lr, [sp, #-16]!
+
+    // Allocate memory for switcher
+    mov       x0, xzr    // address
+                         // length - already stored in `x1`
+    mov       w2, #3     // prot == PROT_READ | PROT_WRITE
+    mov       w3, #4098  // flags == MAP_PRIVATE | MAP_ANONYMOUS
+    mov       w4, #-1    // fd
+    mov       w5, wzr    // offset
+    bl        mmap
+
+    // Restore length and `lr`
+    ldp       x1, lr, [sp], #16
+
+    // Save pointer to new allocated memory in `switcher_caps`
+    adr       x2, switcher_caps
+    str       x0, [x2]
+
+    // Derive DDC for switcher
+    cvtp      c2, x0
+    scbnds    c2, c2, x1
+
+    // Derive PCC for `switch_compartments` and friends
+    adr       x3, switcher_entry
+    adr       x4, switch_compartment_end
+    sub       x4, x4, x3
+    cvtp      c3, x3
+    scbndse   c3, c3, x4
+
+    // Store (DDC, PCC) at `switcher_caps`
+    ldr       x1, switcher_caps
+    stp       c2, c3, [x1], #32
+
+    // Save start address for compartment capabilities in `comps_addr`
+    adr       x2, comps_addr
+    str       x1, [x2]
+
+    ret
+
+/**
+ * Function to add information for a compartment
+ *
+ * @param x0 Compartment memory size
+ * @param x1 Compartment executable function
+ *
+ * @return Pointer to newly allocated memory region
+ */
+.type add_compartment, "function"
+add_compartment:
+    // Store inputs and `lr` so we can call `mmap`
+    stp       x0, x1, [sp, #-32]!
+    str       lr, [sp, #16]
+    mov       x1, x0
+
+    // Allocate memory for new compartment
+    mov       x0, xzr    // address
+                         // length - already stored in `x1`
+    mov       w2, #3     // prot == PROT_READ | PROT_WRITE
+    mov       w3, #4098  // flags == MAP_PRIVATE | MAP_ANONYMOUS
+    mov       w4, #-1    // fd
+    mov       w5, wzr    // offset
+    bl        mmap
+
+    // Restore memory size and function address
+    ldp       x1, x2, [sp], #32
+    ldr       lr, [sp, #-16]
+
+    // Derive compartment DDC
+    cvtp      c0, x0
+    scbnds    c0, c0, x1
+
+    // Derive compartment PCC
+    cvtp      c1, x2
+    mov       x2, #320 // TODO dynamic value
+    scbndse   c1, c1, x2
+
+    // Store new PCC and DDC
+    ldr       x2, comps_addr
+    ldr       x3, comps_cnt
+    mov       x4, #COMP_SIZE
+    madd      x2, x3, x4, x2
+    stp       c0, c1, [x2]
+
+    // Increment counter
+    adr       x3, comps_cnt
+    ldr       x4, [x3]
+    add       x4, x4, #1
+    str       x4, [x3]
+
+    // Update switcher DDC
```

Comment:
Ah right. Initially I was thinking if the switcher DDC should be dynamic, meaning that as we create compartments, therefore associated capabilities, we lengthen the visible space for the switcher (and vice-versa for deleting). But then I thought that we allocate some maximum memory at the start anyway (and currently there is no handling for what happens if we need more memory than that max), so wouldn't dynamically changing the DDC just add complexity, without much gain?

Reply:
Good point. In the longer term, we will need to allow the switcher to create (and delete!) an arbitrary number of compartments. However, that shouldn't change the external API, so it's fine if we only allow a fixed number at first (but if we do that, we should add a check so that we abort the process if the user tries to exceed that fixed number).

Reply:
Implemented in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/61b9d8aedebd12dc9032854df65191f54ca1cc32.

Reply:
Looks good to me -- but can we also delete the commented out code if it's not needed?

Reply:
Done in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/7e163169774b6b5844459b33b570cc87c40d3cc4 for now. I guess we can keep the idea in mind, if we do want to take that approach.

Reply:
Lots of new code seems to have crept in with https://github.com/capablevms/cheri-morello-compartmentalisation/commit/7e163169774b6b5844459b33b570cc87c40d3cc4 ?